File: sdk/go/binary_test.go
package shojiku

import (
	"errors"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// A binary that exists and can be run, for the discovery tests. Its contents
// never matter: nothing here runs it.
func executableAt(t *testing.T, dir, name string) string {
	t.Helper()
	path := filepath.Join(dir, name)
	if err := os.WriteFile(path, []byte("#!/bin/sh\nexit 0\n"), 0o755); err != nil {
		t.Fatalf("writing %s: %v", path, err)
	}
	return path
}

func TestTheEnvironmentBeatsExplicitConfiguration(t *testing.T) {
	dir := t.TempDir()
	fromEnv := executableAt(t, dir, "shojiku")
	other := executableAt(t, dir, "configured")

	bin, err := newBinary(other, &env{enabled: true, source: []string{"SHOJIKU_BIN=" + fromEnv}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if bin.path != fromEnv {
		t.Errorf("path = %q, want the environment's %q", bin.path, fromEnv)
	}
	if bin.source != "environment" {
		t.Errorf("source = %q, want %q", bin.source, "environment")
	}
}

func TestExplicitConfigurationBeatsThePath(t *testing.T) {
	dir := t.TempDir()
	configured := executableAt(t, dir, "configured")
	executableAt(t, dir, "shojiku")

	bin, err := newBinary(configured, &env{enabled: true, source: []string{"PATH=" + dir}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if bin.path != configured || bin.source != "configuration" {
		t.Errorf("got %q from %q, want %q from configuration", bin.path, bin.source, configured)
	}
}

func TestThePathIsSearchedWhenNothingElseNamesOne(t *testing.T) {
	dir := t.TempDir()
	onPath := executableAt(t, dir, "shojiku")

	bin, err := newBinary("", &env{enabled: true, source: []string{"PATH=" + dir}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if bin.path != onPath || bin.source != "path" {
		t.Errorf("got %q from %q, want %q from path", bin.path, bin.source, onPath)
	}
}

func TestTheWindowsSpellingIsTriedAtEveryPathEntry(t *testing.T) {
	empty, holding := t.TempDir(), t.TempDir()
	windows := executableAt(t, holding, "shojiku.exe")

	bin, err := newBinary("", &env{enabled: true, source: []string{
		"PATH=" + empty + string(os.PathListSeparator) + holding,
	}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if bin.path != windows {
		t.Errorf("path = %q, want the .exe spelling %q", bin.path, windows)
	}
}

func TestAnEmptyPathEntryIsSkippedRatherThanSearchedAsTheCwd(t *testing.T) {
	// A leading separator is an empty entry, which POSIX shells read as the
	// current directory. Searching it would make discovery depend on where
	// the process happens to be standing.
	dir := t.TempDir()
	found := executableAt(t, dir, "shojiku")

	bin, err := newBinary("", &env{enabled: true, source: []string{
		"PATH=" + string(os.PathListSeparator) + dir,
	}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if bin.path != found {
		t.Errorf("path = %q, want %q", bin.path, found)
	}
}

func TestNoBinaryAnywhereNamesTheInstallChannels(t *testing.T) {
	_, err := newBinary("", &env{enabled: true, source: []string{"PATH=" + t.TempDir()}})

	if !errors.Is(err, ErrBinaryNotFound) {
		t.Fatalf("err = %v, want ErrBinaryNotFound", err)
	}
	for _, channel := range []string{
		"build it from a repository clone",
		"SHOJIKU_BIN",
		"shojiku.WithBinary",
	} {
		if !strings.Contains(err.Error(), channel) {
			t.Errorf("the message does not name the %q channel: %s", channel, err)
		}
	}
}

func TestAPathThatIsNotExecutableIsRefusedByName(t *testing.T) {
	path := filepath.Join(t.TempDir(), "not-executable")
	if err := os.WriteFile(path, []byte("text"), 0o644); err != nil {
		t.Fatalf("writing the fixture: %v", err)
	}

	_, err := newBinary(path, &env{enabled: true, source: []string{}})

	if !errors.Is(err, ErrBinaryNotFound) {
		t.Fatalf("err = %v, want ErrBinaryNotFound", err)
	}
	if !strings.Contains(err.Error(), "is not an executable file") {
		t.Errorf("the message does not say why: %s", err)
	}
}

func TestDiscoveryDefaultsToTheProcessEnvironment(t *testing.T) {
	// The gate image sets SHOJIKU_BIN, so a lookup with no injected source
	// finds it — which is what proves `source: nil` reads the real thing
	// rather than nothing.
	binary := engineBinary(t)

	bin, err := newBinary("", &env{enabled: true})

	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if bin.path != binary {
		t.Errorf("path = %q, want the process's SHOJIKU_BIN %q", bin.path, binary)
	}
}

func TestAWindowsCandidateNeedsNoExecuteBit(t *testing.T) {
	// The half of the rule the gate's own platform can never exercise:
	// os.Stat reports no execute bits on Windows, so a mode test alone would
	// reject every candidate there.
	readOnly := fs.FileMode(0o444)

	if executableOn(readOnly, true) != true {
		t.Error("a regular file is not executable on windows")
	}
	if executableOn(readOnly, false) != false {
		t.Error("a file with no execute bit is executable off windows")
	}
	if executableOn(fs.ModeDir|0o755, true) != false {
		t.Error("a directory is executable")
	}
}

File: sdk/php/tests/BinaryTest.php
<?php

declare(strict_types=1);

namespace Shojiku\Tests;

use PHPUnit\Framework\TestCase;
use Shojiku\Binary;
use Shojiku\Env;
use Shojiku\Exception\BinaryNotFoundException;

/**
 * Finding the engine, at all three lookup positions.
 *
 * The order is the REVERSE of the template root's, and deliberately: where
 * the engine lives is an operator's decision that has to be able to win over
 * application code, while which templates an application renders is the
 * application's own.
 */
final class BinaryTest extends TestCase
{
    use EngineFixtures;
    use StubBinary;

    public function testTheEnvironmentBeatsExplicitConfiguration(): void
    {
        $this->inTempDir(function (string $dir): void {
            $fromEnv = $this->stubBinary($dir, 'exit 0');
            $env = new Env(enabled: true, source: ['SHOJIKU_BIN' => $fromEnv]);

            $binary = new Binary('/configured/shojiku', $env);

            self::assertSame($fromEnv, $binary->path);
            self::assertSame('environment', $binary->source);
        });
    }

    public function testExplicitConfigurationBeatsThePath(): void
    {
        $this->inTempDir(function (string $dir): void {
            $configured = $this->stubBinary($dir, 'exit 0');
            $onPath = $dir.'/path-copy';
            copy($configured, $onPath);
            chmod($onPath, 0o755);
            $env = new Env(enabled: true, source: ['PATH' => $dir]);

            $binary = new Binary($configured, $env);

            self::assertSame($configured, $binary->path);
            self::assertSame('configuration', $binary->source);
        });
    }

    public function testThePathIsSearchedWhenNothingElseNamesOne(): void
    {
        $this->inTempDir(function (string $dir): void {
            $onPath = $dir.'/shojiku';
            copy($this->engineBinary(), $onPath);
            chmod($onPath, 0o755);
            $env = new Env(enabled: true, source: ['PATH' => '/nowhere'.PATH_SEPARATOR.$dir]);

            $binary = new Binary(null, $env);

            self::assertSame($onPath, $binary->path);
            self::assertSame('path', $binary->source);
        });
    }

    public function testTheWindowsSpellingIsTriedAtEveryPathEntry(): void
    {
        // Windows is a first-class target, and a lookup that only tried the
        // bare name would fail there alone — the platform least likely to be
        // in front of whoever writes the lookup.
        $this->inTempDir(function (string $dir): void {
            $onPath = $dir.'/shojiku.exe';
            file_put_contents($onPath, "#!/bin/sh\nexit 0\n");
            chmod($onPath, 0o755);
            $env = new Env(enabled: true, source: ['PATH' => $dir]);

            self::assertSame($onPath, (new Binary(null, $env))->path);
        });
    }

    public function testAnEmptyPathEntryIsSkippedRatherThanSearchedAsTheCwd(): void
    {
        $env = new Env(enabled: true, source: ['PATH' => PATH_SEPARATOR]);

        self::assertSame([null, 'path'], Binary::discover(null, $env));
    }

    public function testNoBinaryAnywhereNamesTheInstallChannels(): void
    {
        $env = new Env(enabled: true, source: ['PATH' => '/nowhere']);

        try {
            new Binary(null, $env);
            self::fail('a missing binary was not refused');
        } catch (BinaryNotFoundException $e) {
            self::assertStringContainsString('never downloads the engine', $e->getMessage());
            self::assertStringContainsString('SHOJIKU_BIN', $e->getMessage());
            self::assertStringContainsString('Docker image', $e->getMessage());
        }
    }

    public function testAPathThatIsNotExecutableIsRefusedByName(): void
    {
        $this->inTempDir(function (string $dir): void {
            // A DIRECTORY where the binary belongs, not a chmod: the gate
            // container runs as root, so permission bits prove nothing.
            mkdir($dir.'/shojiku');
            $env = new Env(enabled: true, source: []);

            $this->expectException(BinaryNotFoundException::class);
            $this->expectExceptionMessageMatches('/is not an executable file/');
            new Binary($dir.'/shojiku', $env);
        });
    }

    public function testDiscoveryDefaultsToTheProcessEnvironment(): void
    {
        // The image sets SHOJIKU_BIN, which is also how every other test in
        // this suite reaches the engine.
        self::assertSame($this->engineBinary(), (new Binary())->path);
    }
}
